### ISSUE TITLE
Limitations on conditionals in phase 1 of initializers

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#8075]
While writing some initializers this week, I wrote some of the form:

```chapel
proc init(flag: bool) {
  if flag then
    this.x = ...;
}
```

and got the error message:

```
error: cannot initialize fields in an if statement in phase 1
```

Yet, if I wrote the conditional:

```chapel
proc init(flag: bool) {
  if flag then
    this.x = ...;
  else
    this.x = ...;
}
```

that was OK.  This led to a series of thoughts:

1) If we want to distinguish between these two cases, then the error message should be made more specific since the second example demonstrates that it is possible to initialize fields in a conditional (better term than "if statement" btw), but that that if statement must do so in both branches.  So perhaps "cannot initialize fields in just one branch of a conditional in phase 1"?

2) That said, it also made me wonder what prevented the compiler from transforming:

```chapel
  if flag then
    this.x = ...;
```

into:

```chapel
  if flag then
    this.x = ...;
  else
    this.x = /* default initialization that would be inserted if there had been no conditional */;
```

3) Finally, in my original case, the `flag` was actually a `param` boolean, and it seemed as though the compiler could/should flatten the conditional and understand that the field was either always set or always not set as though it had not been located within a conditional.  That is, it seemed as though it ought to be legal even if we did not pursue idea 2.

For context, I ran into these while converting the range constructors over to initializers...  In this case, there are other, reasonable, ways to work around the issue, but I thought I'd note these thoughts before moving on and forgetting about them.

**Comment by lydia-duncan** [comment]
I agree with 1 (including that we should use "conditional" instead of "if statement", though I don't recall off-hand what we do about select statements).  I think 2 is a reasonable extension (though likely lower priority than actual bugs).

For 3, I agree, but our validation for initializers by necessity runs prior to our param handling (normalization butchers all these statements on its own, to the point where they are unrecognizable without some serious compiler fortune telling).  I think this is something that will either require reimplementing param handling earlier in compilation, or require a reorganization of the compiler.

**Comment by bradcray** [comment]
I figured pass ordering was going to be a problem for 3.  Just so I understand the constraints, what is the "by necessity" that requires it to run before param folding?  Thanks.

**Comment by lydia-duncan** [comment]
It's really about the ordering of normalization relative to resolution (when param folding occurs).  Things like method and function calls and even other field accesses get separated from the field that would store that result, making it a lot more effort to determine what is happening to a particular field.  The whole picture of a field assignment will be spread across multiple CallExprs at resolution, rather than a single one prior to normalization.

**Comment by benharsh** [comment]
To clarify, I think Items 1 and 2 are complete, but 3 (param conditionals) are still not treated specially.

**Comment by cassella** [comment]
All the variants of this I could think of from what was described work for me, including item 3:


```chapel
record R {
  var x: int;
  proc init(flag: bool) {
    if flag then
      this.x = 1;
  }
}

record S {
  var x: int;
  proc init(param flag: bool) {
    if flag then
      this.x = 1;
  }
}


var r1 = new R(true);
var r2 = new R(false);

var s1 = new S(true);
var s2 = new S(false);

writeln(r1);
writeln(r2);

writeln(s1);
writeln(s2);
```

```
(x = 1)
(x = 0)
(x = 1)
(x = 0)
```

Is there anything remaining to do here?

**Comment by lydia-duncan** [comment]
It seems reasonable to me to close this, but I'm not seeing a param conditional test, would you be willing to open a PR with yours and I'll merge it?

**Comment by lydia-duncan** [comment]
Merged test from Paul locking in that the last outstanding portion of this issue is resolved, closing
